# 导入模块

In [1]:
import pandas as pd
from pyecharts.charts import *
from pyecharts import options as opts
from pyecharts.globals import ThemeType
from pyecharts.globals import SymbolType
from pyecharts.commons.utils import JsCode
from pyecharts.charts import Line,Bar,Page,Map

# 检查数据

## 结婚登记（万对）数据

In [2]:
df_marry = pd.read_excel('结婚登记(万对).xlsx')
df_marry.head()

地区  2019年  2018年  2017年  2016年  2015年  2014年  2013年  2012年  2011年  \
0     北京市  12.90  13.78  15.15  16.62  16.60  17.00  16.37  17.41  17.32   
1     天津市   9.64   9.75   9.51   9.82  10.12   9.94  10.26  10.14  10.41   
2     河北省  42.13  45.87  50.49  55.19  60.94  66.13  74.08  74.53  77.72   
3     山西省  25.50  27.85  28.77  30.01  34.68  35.07  38.40  36.28  33.96   
4  内蒙古自治区  16.18  17.69  18.70  19.84  21.79  21.68  22.16  20.77  21.53   

   2010年  2009年  2008年  2007年  2006年  2005年  2004年  2003年  2002年  2001年  
0   13.8  18.18  14.75  11.79   17.1    9.7  12.64    9.4    7.6   7.94  
1    8.7  10.40   8.93   7.57    8.4    6.4   7.70    6.3    5.4   5.79  
2   75.0  72.00  66.31  60.39   55.5   53.8  57.48   52.4   47.1  44.56  
3   36.1  34.36  28.74  23.85   20.2   19.0  16.15   14.5   15.4  14.62  
4   20.3  18.84  18.78  18.57   16.3   15.5  14.77   13.1   13.1  14.76

## 离婚登记（万对）数据

In [3]:
df_divorce = pd.read_excel('离婚登记(万对).xlsx')
df_divorce.head()

地区  2019年  2018年  2017年  2016年  2015年  2014年  2013年  2012年  2011年  \
0     北京市   8.38   7.41   8.06  10.58   8.22   6.56   6.46   4.86   4.35   
1     天津市   7.47   6.41   5.89   6.52   5.19   4.45   4.53   3.63   3.21   
2     河北省  25.57  23.45  23.24  22.02  19.90  18.57  17.88  17.12  15.63   
3     山西省   9.12   8.75   8.22   7.65   7.22   6.56   6.10   5.28   4.75   
4  内蒙古自治区   9.98   9.88  10.09   9.84   9.19   8.84   8.07   7.06   6.67   

   2010年  2009年  2008年  2007年  2006年  2005年  2004年  2003年  2002年  2001年  
0    4.4   4.13   3.76   3.66    3.6    3.4    3.3    3.1    2.8   2.77  
1    3.0   2.76   2.40   2.21    2.1    1.9    1.8    1.4    1.2   1.31  
2   14.5  12.67  11.22   9.88    9.2    8.3    7.7    5.8    5.3   5.16  
3    4.3   3.89   3.28   3.24    3.0    2.9    2.7    2.3    2.3   2.29  
4    5.7   5.20   5.02   4.73    4.1    3.9    3.7    3.0    2.5   3.31

## 检查数据完整度

In [4]:
df_marry.isnull().any()

地区       False
2019年    False
2018年    False
2017年    False
2016年    False
2015年    False
2014年    False
2013年    False
2012年    False
2011年    False
2010年    False
2009年    False
2008年    False
2007年    False
2006年    False
2005年    False
2004年    False
2003年    False
2002年    False
2001年    False
dtype: bool

In [5]:
df_divorce.isnull().any()

地区       False
2019年    False
2018年    False
2017年    False
2016年    False
2015年    False
2014年    False
2013年    False
2012年    False
2011年    False
2010年    False
2009年    False
2008年    False
2007年    False
2006年    False
2005年    False
2004年    False
2003年    False
2002年    False
2001年    False
dtype: bool

# 新增列：结婚占比、离婚占比、离结率-2019

In [6]:
# 2019年各地区离结率
df_tmp = pd.DataFrame()
df_tmp['地区'] = df_marry['地区']
df_tmp['结婚'] = df_marry['2019年']
df_tmp['离婚'] = df_divorce['2019年']
df_tmp['结婚占比'] = round(df_marry['2019年']*100 /(df_marry['2019年'] + df_divorce['2019年']), 2)
df_tmp['离婚占比'] = round(df_divorce['2019年']*100 /(df_marry['2019年'] + df_divorce['2019年']), 2)
df_tmp['离结率'] = round(df_tmp['离婚占比'] /(df_tmp['结婚占比']), 2)
df_tmp.head()

地区     结婚     离婚   结婚占比   离婚占比   离结率
0     北京市  12.90   8.38  60.62  39.38  0.65
1     天津市   9.64   7.47  56.34  43.66  0.77
2     河北省  42.13  25.57  62.23  37.77  0.61
3     山西省  25.50   9.12  73.66  26.34  0.36
4  内蒙古自治区  16.18   9.98  61.85  38.15  0.62

# 可视化

## 2019年各地区结婚离婚占比

In [7]:
# 阴影样式
itemstyle={
    'normal': {
        'shadowColor': 'rgba(0, 0, 0, .5)',  # 阴影颜色
        'shadowBlur': 5,  # 阴影大小
        'shadowOffsetY': 2,  # Y轴方向阴影偏移
        'shadowOffsetX': 2,  # x轴方向阴影偏移
        'borderColor': '#fff'
    }
}

area = df_tmp['地区'].values.tolist()
marry_count = df_tmp['结婚占比'].values.tolist()
divorce_count = df_tmp['离婚占比'].values.tolist()

b1 = (
    Bar(
        init_opts=opts.InitOpts(
        width='800px', height='600px',
        )
    )
    .add_xaxis(area)
    .add_yaxis('结婚占比', marry_count, stack='stack1',itemstyle_opts=opts.ItemStyleOpts(color='#FF8066'))
    .add_yaxis('离婚占比', divorce_count, stack='stack1',itemstyle_opts=opts.ItemStyleOpts(color='#B0A8B9'))
    .set_series_opts(
        label_opts=opts.LabelOpts(
             is_show=True,
             position='inside'
        )
    )
    .set_global_opts(
        legend_opts=opts.LegendOpts(
            pos_right='10%',
            pos_top='2%',
            orient='horizontal',
        ),
        title_opts=opts.TitleOpts(
            title='2019年各地区结婚离婚占比',
            pos_top='2%'
        ),
    )
    .reversal_axis()
)
b1.render('a.html')

'C:\\Users\\19108\\Desktop\\Internet _New Media\\交互式数据可视化\\final\\finaltest\\a.html'

## 2019离结率又高至低排序高亮图

In [8]:
df_tmp = df_tmp.sort_values('离结率', ascending=False)
df_tmp.style.bar(subset=['离结率'],color='#CD3F2C')

## 全国离结率TOP16渐变柱状图

In [9]:
sort_info = df_tmp.sort_values('离结率', ascending=True)
b1 = (
    Bar(init_opts=opts.InitOpts(
        width='800px', height='600px',))
    .add_xaxis(list(sort_info['地区'])[-16:])
    .add_yaxis('', sort_info['离结率'].values.tolist()[-16:],
               category_gap='30%',
               itemstyle_opts={
               'normal': {
                   'shadowColor': 'rgba(0, 0, 0, .5)',  # 阴影颜色
                   'shadowBlur': 5,  # 阴影大小
                   'shadowOffsetY': 2,  # Y轴方向阴影偏移
                   'shadowOffsetX': 2,  # x轴方向阴影偏移
                   'borderColor': '#fff'
                   }
               }
              )
    .reversal_axis()
    .set_global_opts(
        xaxis_opts=opts.AxisOpts(is_show=False),    
        yaxis_opts=opts.AxisOpts(is_show=False,
            axisline_opts=opts.AxisLineOpts(is_show=False),
            axistick_opts=opts.AxisTickOpts(is_show=False)
        ),
        title_opts=opts.TitleOpts(
            title='全国离结率 TOP16',
            pos_left='9%',
            pos_top='4%',
            title_textstyle_opts=opts.TextStyleOpts(
                color='#ed1941', font_size=16)
        ),
        visualmap_opts=opts.VisualMapOpts(
            is_show=False,
            max_=15,
            series_index=0,
        ),
     )
    .set_series_opts(
        itemstyle_opts={
            "normal": {
                "color": JsCode(
                    """new echarts.graphic.LinearGradient(0, 0, 0, 1, [{
                offset: 0,
                color: '#ed1941'
            }, {
                offset: 1,
                color: '#009ad6'
            }], false)"""
                ),
                "barBorderRadius": [30, 30, 30, 30],
                "shadowColor": "rgb(0, 160, 221)",
            }
        },
        label_opts=opts.LabelOpts(position="insideLeft",
                                              font_size=10,
                                              vertical_align='middle',
                                              horizontal_align='left',
                                              font_weight='bold',
                                              formatter='{b}: {c}'))
)

b1.render('b.html')

'C:\\Users\\19108\\Desktop\\Internet _New Media\\交互式数据可视化\\final\\finaltest\\b.html'

In [10]:
print('全国共有{}个省市地区离结率超过50%，这已经是相当高了'.format(len(df_tmp[df_tmp['离结率'] > 0.5])))

全国共有15个省市地区离结率超过50%，这已经是相当高了


In [11]:
df_tmp[df_tmp['离结率'] > 0.5]['地区']

1        天津市
7       黑龙江省
6        吉林省
5        辽宁省
21       重庆市
0        北京市
8        上海市
4     内蒙古自治区
2        河北省
17       湖南省
16       湖北省
9        江苏省
10       浙江省
14       山东省
22       四川省
Name: 地区, dtype: object

## 广东历年结婚/离婚登记数据

In [12]:
# 线性渐变
color_js0 = """new echarts.graphic.LinearGradient(0, 1, 0, 0,
    [{offset: 0, color: '#FFFFFF'}, {offset: 1, color: '#ed1941'}], false)"""

color_js1 = """new echarts.graphic.LinearGradient(0, 1, 0, 0,
    [{offset: 0, color: '#FFFFFF'}, {offset: 1, color: '#009ad6'}], false)"""


city = '广东省'
b2 = (
    Bar()
    .add_xaxis(df_divorce.columns[1:].values.tolist())
    .add_yaxis(
        series_name="结婚登记",
        y_axis=df_marry[df_marry['地区']==city].values[0][1:].tolist(),
        markpoint_opts=opts.MarkPointOpts(
            data=[
                opts.MarkPointItem(type_="max", name="最大值"),
                opts.MarkPointItem(type_="min", name="最小值"),
            ]
        ),
        itemstyle_opts=opts.ItemStyleOpts(color=JsCode(color_js0))
    )
    .add_yaxis(
        series_name="离婚登记",
        y_axis=df_divorce[df_divorce['地区']=='广东省'].values[0][1:].tolist(),
        markpoint_opts=opts.MarkPointOpts(
            data=[
                opts.MarkPointItem(type_="max", name="最大值"),
                opts.MarkPointItem(type_="min", name="最小值"),
            ]
        ),
        itemstyle_opts=opts.ItemStyleOpts(color=JsCode(color_js1))
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(title=""),
        tooltip_opts=opts.TooltipOpts(trigger="axis"),
        xaxis_opts=opts.AxisOpts(name="年份",type_="category", 
                                 boundary_gap=True,
                                 axisline_opts=opts.AxisLineOpts(is_show=True,
                                                                 linestyle_opts=opts.LineStyleOpts(width=4, color='#DB7093')),
                                 axislabel_opts=opts.LabelOpts(rotate=45)),
        yaxis_opts=opts.AxisOpts(
                axislabel_opts=opts.LabelOpts(formatter="{value} /万对"),
                name=f'{city}历年结婚/离婚登记',            
                is_scale=True,
                name_textstyle_opts=opts.TextStyleOpts(font_size=15,font_weight='bold',color='#FF1493'),
                splitline_opts=opts.SplitLineOpts(is_show=True, 
                                                  linestyle_opts=opts.LineStyleOpts(type_='dashed')),
                axisline_opts=opts.AxisLineOpts(is_show=False,
                                        linestyle_opts=opts.LineStyleOpts(width=2, color='#DB7093'))
            ),
        legend_opts=opts.LegendOpts(is_show=True, pos_top='2%', legend_icon='roundRect'),
    )
)
b2.render('c.html')

'C:\\Users\\19108\\Desktop\\Internet _New Media\\交互式数据可视化\\final\\finaltest\\c.html'

由图我们可以看出广东的离婚率几乎是逐年稳步提升，结婚率近五年来逐年下降

## 2019年各地区离结率数据热力图

In [13]:
# 线性渐变
color_js = """new echarts.graphic.LinearGradient(0, 0, 1, 0,
    [{offset: 0, color: '#009ad6'}, {offset: 1, color: '#ed1941'}], false)"""
df_tmp = df_tmp.replace('省', '', regex=True).replace('市', '', regex=True).replace('自治区', '', regex=True).replace('壮族', '', regex=True).replace('维吾尔', '', regex=True).replace('回族', '', regex=True)
map_chart = Map(init_opts=opts.InitOpts(theme='light',
                                        width='800px',
                                        height='600px'))
map_chart.add('离婚/结婚',
              [list(z) for z in zip(df_tmp['地区'].values.tolist(), df_tmp['离结率'].values.tolist())],
              maptype='china',
              is_map_symbol_show=False,
              itemstyle_opts={
                  'normal': {
                      'shadowColor': 'rgba(0, 0, 0, .5)',  # 阴影颜色
                      'shadowBlur': 5,  # 阴影大小
                      'shadowOffsetY': 0,  # Y轴方向阴影偏移
                      'shadowOffsetX': 0,  # x轴方向阴影偏移
                      'borderColor': '#fff'
                  }
              }
              )
map_chart.set_global_opts(
    visualmap_opts=opts.VisualMapOpts(
        is_show=True,
        is_piecewise=True,
        min_ = 0,
        max_ = 1,
        split_number = 5,
        series_index=0,
        pos_top='70%',
        pos_left='10%',
        range_text=['离结率', ''],
        pieces=[
            {'max':1.0, 'min':0.8, 'label':'0.8-1.0', 'color': '#990000'},
            {'max':0.8, 'min':0.6, 'label':'0.6-0.8', 'color': '#CD5C5C'},
            {'max':0.6, 'min':0.4, 'label':'0.4-0.6', 'color': '#F08080'},
            {'max':0.4, 'min':0.2, 'label':'0.2-0.4', 'color': '#FFCC99'},
            {'max':0.2, 'min':0.0, 'label':'0.0-0.2', 'color': '#FFE4E1'},
           ],
    ),
    legend_opts=opts.LegendOpts(is_show=False),
    tooltip_opts=opts.TooltipOpts(
        is_show=True,
        trigger='item',
        formatter='{b}:{c}'
    ),
    title_opts=dict(
        text='2019年各地区离结率',
        left='center',
        top='5%',
        textStyle=dict(
            color='#DC143C'))
)
map_chart.render_notebook()

## 历年全国结婚离婚登记趋势

In [14]:
df_marry.loc["总计"] =df_marry.apply(lambda x:x.sum()) 
df_divorce.loc["总计"] =df_divorce.apply(lambda x:x.sum()) 
df_marry.loc['总计','地区'] = '全国'
df_divorce.loc['总计','地区'] = '全国'

In [15]:
df_marry[df_marry['地区'] == '全国']

地区   2019年    2018年   2017年    2016年    2015年    2014年    2013年    2012年  \
总计  全国  927.33  1013.94  1063.1  1142.84  1224.71  1306.72  1346.93  1323.58   

      2011年   2010年    2009年   2008年   2007年  2006年  2005年   2004年   2003年  \
总计  1302.35  1241.2  1212.43  1098.3  991.38  944.9  823.1  867.21  810.64   

    2002年   2001年  
总计  786.1  804.95

In [16]:
df_divorce[df_marry['地区'] == '全国']

地区   2019年   2018年   2017年   2016年   2015年   2014年   2013年   2012年  \
总计  全国  470.09  446.07  437.35  415.81  384.14  363.65  350.02  310.42   

     2011年  2010年   2009年   2008年   2007年  2006年  2005年  2004年  2003年  2002年  \
总计  287.38  267.8  246.78  226.89  209.83  191.0  178.4  166.4  132.4  117.6   

     2001年  
总计  125.07

In [18]:
# 背景色
background_color_js = """
new echarts.graphic.RadialGradient(0.3, 0.3, 0.8, [{offset: 0,color: '#FEFEDF'}, 
{offset: 1,color: '#FEF7FF'}])
"""
 
# 线条样式
linestyle_dic = { 'normal': {
                    'width': 4,  
                    'shadowColor': '#696969', 
                    'shadowBlur': 10,  
                    'shadowOffsetY': 10,  
                    'shadowOffsetX': 10,  
                    }
                }
l1 = (
    Line(init_opts=opts.InitOpts(bg_color=JsCode(background_color_js)))
    .add_xaxis(xaxis_data=df_divorce.columns[1:][::-1])
    .add_yaxis(
        series_name="结婚登记",
        y_axis=[round(i,2) for i in df_marry.loc["总计"].values.tolist()[1:]],
        symbol_size=8,
        is_smooth=True,
        color="#009ad6",
    )
    .add_yaxis(
        series_name="离婚登记",
        y_axis=[round(i,2) for i in df_divorce.loc["总计"].values.tolist()[1:]],
        symbol_size=8,
        is_smooth=True,        
        color="#ed1941",
    )
    # 系列配置项
    .set_series_opts(linestyle_opts=linestyle_dic,
                     areastyle_opts=opts.AreaStyleOpts(opacity=0.6),
                     label_opts=opts.LabelOpts(is_show=False),
                     markline_opts=opts.MarkLineOpts(data=[opts.MarkLineItem(type_="average")]),
                     markpoint_opts=opts.MarkPointOpts(
                        data=[opts.MarkPointItem(type_="max"), opts.MarkPointItem(type_="min")],
                        symbol_size=[65, 50],
                        label_opts=opts.LabelOpts(position="inside", color="#fff", font_size=10)
                        ),
                    )
    # 通用配置项
    .set_global_opts(
        title_opts=opts.TitleOpts(
                title='历年全国结/离婚登记趋势',
                pos_top='2%',
                title_textstyle_opts=opts.TextStyleOpts(color='#474554', font_size=20)),
        tooltip_opts=opts.TooltipOpts(trigger="axis"),
        xaxis_opts=opts.AxisOpts(name="",type_="category", 
                         boundary_gap=True,
                         axisline_opts=opts.AxisLineOpts(is_show=True,
                                                         linestyle_opts=opts.LineStyleOpts(width=2, color='#787586')),
                         axislabel_opts=opts.LabelOpts(rotate=45)),
        yaxis_opts=opts.AxisOpts(
                axislabel_opts=opts.LabelOpts(formatter="{value} /万对"),
                is_scale=True,
                name_textstyle_opts=opts.TextStyleOpts(font_size=12,font_weight='bold',color='#474554'),
                splitline_opts=opts.SplitLineOpts(is_show=True, 
                                                  linestyle_opts=opts.LineStyleOpts(type_='dashed')),
                axisline_opts=opts.AxisLineOpts(is_show=False,
                                        linestyle_opts=opts.LineStyleOpts(width=2, color='#787586'))
            ),
        # 图例样式
        legend_opts=opts.LegendOpts(is_show=True, pos_right='1%', pos_top='2%',legend_icon='roundRect'),
    )
)
l1.render_notebook()

# 总结
**从我们统计分析的数据可以得知，我国各省份结婚总对数从2001年至2013年保持上升趋势达到峰值1346.93对，从2014年开始减少，而离婚率从2001年以来持续上升  
1.结婚登记数量前五地区：河南、广东、四川、江苏、安徽  
2.离婚登记数量前五地区：河南、四川、江苏、山东、河北  
3.全国结婚登记数量从2007年开始出现持续下降趋势  
4.北方地区离结率较南方地区高出不少**